grid should be axis attribute (fix #675)

daschw opened the pull request:
This makes `grid` and `foreground_color_grid` Axis attributes instead of Subplot attributes (cf #675).
A lot of this work has already been done by @GuessWhoSamFoo in #695, so credits to him!!!
I just tried to make it work on all different backends and also made `foreground_color_grid` an Axis attribute.

I tested the changes with the following code:
```julia
using Plots

x = rand(10)
plot(
    plot(x, xgrid = false, title = "xgrid = false"),
    plot(x, ygrid = nothing, title = "ygrid = nothing"),
    plot(x, grid = false, title = "grid = false"),
    plot(x, xforeground_color_grid = :red, title = "xfcg = :red"),
    plot(x, foreground_color_grid = :red, title = "fcg = :red"),
    plot(x, xforeground_color_grid = :red, yforeground_color_grid = :green, title = "xfcg = :red, yfcg = :green"),
    label = "",
    )
```

Here are the resulting plots for the different backends:
* **GR**
![grid_axis_gr](https://user-images.githubusercontent.com/16589944/29187939-c14368a6-7e11-11e7-9c70-e7341814a624.png)

* **PyPlot**
![grid_axis_pyplot](https://user-images.githubusercontent.com/16589944/29188033-0e957590-7e12-11e7-9a6d-3593548cadac.png)

* **Plotly(JS)**
![grid_axis_plotlyjs](https://user-images.githubusercontent.com/16589944/29188418-4a7b2c7a-7e13-11e7-8ecd-1561ebfb1b49.png)

* **GLVisualize**
![grid_axis_glvisualize](https://user-images.githubusercontent.com/16589944/29188164-8cafbab2-7e12-11e7-9dfd-561379c8e5e4.png)

The following backends do not support `foreground_color_grid`:

* **InspectDR**
![grid_axis_inspectdr](https://user-images.githubusercontent.com/16589944/29188192-9ff19d3e-7e12-11e7-896e-4efe836f2cab.png)

* **PGFPlots**
![grid_axis_pgfplots](https://user-images.githubusercontent.com/16589944/29188211-b0f8ed08-7e12-11e7-9fba-fb014d6a2a5c.png)

@jheinen, @SimonDanisch and @ma-laforge could you please check the changes to your respective backends?

mkborregaard commented:
Nice! great effort! But what's with the PlotlyJS grid, it looks awful?

daschw commented:
Yes, now that you say it. It was not like this before, right? I'll have a look what I messed up here.

mkborregaard commented:
I'm not sure that's this PR, it looks like that on master as well (on my system) but it didn't always. Compare this from our docs
![skaermbillede 2017-08-11 kl 10 30 15](https://user-images.githubusercontent.com/8429802/29205995-2f04f8d2-7e80-11e7-8c84-2d6191e0ca61.png)
with this from today's master
![skaermbillede 2017-08-11 kl 10 30 25](https://user-images.githubusercontent.com/8429802/29205999-36df06a6-7e80-11e7-9a94-68ce203906fb.png)
I wonder if something went amiss in Plots or a setting was changed in PlotlyJS?

daschw commented:
Weird, this does not happen on my install with plots and plotlyjs master ...

mkborregaard commented:
But it does happen with this (ds-grid-axes) branch?

daschw commented:
Yes, but that's expected (as I just found out).
Here's the issue with PlotlyJS: Before this PR `foreground_color_grid` was not a supported attribute for plotly(js). Thus the plotly default (this very transparent gray) was chosen for the grid lines. The user, however, had no control of the grid colors. With the support of `foreground_color_grid` the grid color now defaults to match `foreground_color_subplot` which is black I think. I see three options here:
* Keep the old behaviour (no `foreground_color_grid` support for plotly(js))
* The current implementation of this branch: Most adjustment options by the user but maybe not the nicest defaults.
* Force an alpha on the grid color like in gr or glvisualize, so the user can choose the color, but there will always be the same alpha value defined by us.

I would go for the 3rd option as a compromise between the first two and try to find out the right alpha value so the plots without grid color set look the same as they do now. But maybe you prefer another option or see another alternative?

mkborregaard commented:
I'd tend to prefer 1 - just keep that as it is for plotlyjs

mkborregaard commented:
BTW, I find I get the ugly grid lines in Juno, but not in the REPL.

daschw commented:
Hm, I don't get it in either.

mkborregaard commented:
weird.

daschw commented:
Ok, Plotly(JS) now looks like this (first option from above - no grid color support):
![grid_axis_plotlyjs_repl](https://user-images.githubusercontent.com/16589944/29210688-f06c8c4c-7e94-11e7-8f95-174c3e61a89f.png)

BTW, now I am also experiencing your issue. The above image is from the REPL when running the same code in Juno an image with all black lines is displayed, and saving it as PNG from Juno results in something inbetween:
![grid_axis_plotlyjs_juno](https://user-images.githubusercontent.com/16589944/29210750-4eda35c2-7e95-11e7-8644-343a699519f2.png)

I don't think however, that this is related to this PR. Still, very strange 😕 !

mkborregaard commented:
It's probably unrelated, yeah.

daschw commented:
One more thing before merging: Currently the (x,y)grid is turned off for the attribute values `false` and `nothing`. Any opinions on whether the `grid` attribute should only accept `Bool` or maybe also act on symbols (and possibly strings) like `:none`,  `:off`, ... ?
A further possibility could be that `grid = :x` or `grid = :y` only shows the grid on the respective axis.

mkborregaard commented:
Yes, I like both of these options. Very much in Plots' spirit.

mkborregaard commented:
In principle, the best way might be to do something different - to make `grid` a magic argument, that can take the `x`, `foreground_color_grid` etc, allowing you to specify `grid = (:x, :red, :dot)`. @tbreloff suggested something similar here https://github.com/JuliaPlots/Plots.jl/pull/721#issuecomment-284759627 (though not exactly the same - I just think a magic argument might be better than asking for `Stroke`). That would be more versatile. 

But that may be overkill?

daschw commented:
I like your suggestions with the magic arguments and being able to set the stroke, but it might take me a few more days. :smile:

mkborregaard commented:
It doesn't have to be in this PR - it could be a separate project after this is merged.

daschw commented:
OK, I implemented the magic argument `grid` now - I hope mostly in Plots style ;-)
I introduced new axis attributes: `gridalpha`, `gridlinewidth` and `gridstyle`, which can be set either directly or by passing a tuple to the grid attribute. E.g: 
```julia
plot(..., grid = (:x, # xgrid = true, ygrid = false
    :solid, # gridstyle
    :blue, # foreground_colour_grid
    0.4, # gridalpha
    2 # gridlinewidth
    ))
```
All grid arguments are supported for GR, PyPlot and GLVisualize.
Plotly(JS) supports all but `gridstyle`. (Note: This leads to a new default look for plotly(js) grid lines unless we do not support foreground_color_grid, see also the discussion above)
I have not implemented the new attributes for InspectDR and PGFPlots.
I added a test example for the magic grid arguments:
```julia
x = rand(10)
p1 = plot(x, title = "Default looks")
p2 = plot(x, grid = (:y, :olivedrab, :solid, 0.5), title = "Modified y grid")
p3 = plot(deepcopy(p2), title = "Add x grid")
xgrid!(:on, :cadetblue, 2, :dashdot)
plot(p1, p2, p3, layout = (1, 3), label = "", fillrange = 0, fillalpha = 0.3)
```
If we want to merge this I will add the respective images to PlotReferenceImages soon.
Here is the testexample for
* **GR**
![magicgrid_gr](https://user-images.githubusercontent.com/16589944/29371915-fb9d91f6-82a9-11e7-8da5-b79353ce4b26.png)

* **PyPlot**
![magicgrid_pyplot](https://user-images.githubusercontent.com/16589944/29371939-0d33ddbc-82aa-11e7-86ce-ce5417375ed5.png)

* **Plotly(JS)**
![magicgrid_plotlyjs](https://user-images.githubusercontent.com/16589944/29371957-1b846c6a-82aa-11e7-919d-44a59de3a2d9.png)

I guess we can still discuss the defaults and the PlotlyJS behaviour.

mkborregaard commented:
Awesome. This is great!
Only two things - WRT plotlyjs I guess there is no way of getting most of this behaviour but keeping the default lines? They are a lot prettier iMHO.
Also just one thing to consider - `grid` has a second meaning in Plots, used to create a regular subplot layout. I don't know if this may cause some confusion?

mkborregaard commented on src/args.jl:
```diff
@@ -163,6 +163,24 @@ const _scaleAliases = Dict{Symbol,Symbol}(
     :log  => :log10,
 )
 
+const _allGridSyms = [:x, :y, :z,
```
nice

mkborregaard commented:
Nice code. This turned out to be a pretty big effort! Apart from my small considerations above, this is ready to go in IMHO.

daschw commented:
Thanks a lot for your feedback Michael!
I have one more suggestion regarding the plotlyjs issue. We could choose different defaults on all backends, so that the plotlyjs default grid looks nicer and the grid attributes can be supported by plotly.
I'd propose something like `gridstyle = :solid, gridalpha = 0.2, gridlinewidth = 0.5, foreground_color_grid = :match`:
* **GR**
![grid_defaults_gr](https://user-images.githubusercontent.com/16589944/29401663-665ff4ec-8332-11e7-84d8-55e4af87fdbf.png)
* **PyPlot**
![grid_defaults_pyplot](https://user-images.githubusercontent.com/16589944/29401682-7dc7fa6c-8332-11e7-8d93-c8e4347116a3.png)
* **PlotlyJS**
![grid_defaults_plotly](https://user-images.githubusercontent.com/16589944/29401696-89ab5784-8332-11e7-8eeb-980b93a7fc95.png)

If you think this is too big a change to the default looks of GR and PyPlot, I am fine with not supporting the grid attributes in plotlyjs as well.

mkborregaard commented:
In that case I'd prefer the stippled-style grid lines. I am a bit on the fence here, but leaning towards putting full control in Plots, overriding the defaults for all the backends, and using something very much like you last example but stippled lines. But I am not really sure...

daschw commented:
Hm `gridstyle = ?`
* **GR**
![grid_defaults_gr_style](https://user-images.githubusercontent.com/16589944/29403141-af1d1106-8337-11e7-841c-0aa1cc6aa1ef.png)

* **PyPlot**
![grid_defaults_pyplot_style](https://user-images.githubusercontent.com/16589944/29403157-b91e7f50-8337-11e7-862a-446e0ec19cc2.png)

* **PlotlyJS** does not support gridstyle so it's the same as above.

We can also ask others for their opinion. Either the Plots members, or in the Gitter channel.

mkborregaard commented:
gridstyle = dot 👍 
Def nice with some more views.

daschw commented:
Does this also close #296?

mkborregaard commented:
Oh, nice old issue :-) @pkofod ?

pkofod commented:
I guess it pretty much does, and if it doesn't completely, then new, specific issues can be opened !
